astropy/modeling/fitting.py

```diff
@@ -1279,16 +1279,25 @@ def _filter_non_finite(self, x, y, z=None, weights=None):
         MESSAGE = "Non-Finite input data has been removed by the fitter."
 
         if z is None:
-            mask = np.isfinite(y)
+            mask = np.isfinite(y) & (
+                np.isfinite(weights)
+                if weights is not None
+                else np.ones_like(y).astype(bool)
```

Comment:
This and the essentially identical statement below can be replaced with

```python
mask = np.ones_like(x, dtype=bool) if weights is None else np.isfinite(weights)
```

and an in-place `&` operation.

Indeed this function can be replaced with:
```python
    def _filter_non_finite(self, x, y, z=None, weights=None):
        """
        Filter out non-finite values in x, y, z.

        Returns
        -------
        x, y, z : ndarrays
            x, y, and z with non-finite values filtered out.
        """
        MESSAGE = "Non-Finite input data has been removed by the fitter."

        mask = np.ones_like(x, dtype=bool) if weights is None else np.isfinite(weights)
        mask &= np.isfinite(y) if z is None else np.isfinite(z)

        if not np.all(mask):
            warnings.warn(MESSAGE, AstropyUserWarning)

        return x[mask], y[mask], None if z is None else z[mask], None if weights is None else weights[mask]
```

Reply:
Thanks, revised in https://github.com/astropy/astropy/pull/15215/commits/13b331a4dc6cd9bdc3b197da958babdb1c72587a.